src/infra/core/src/ingest/push_ingest_service_impl.rs

```diff
@@ -0,0 +1,368 @@
+        )
+    )]
+    async fn do_ingest_inner(
+        &self,
+        mut args: PushIngestArgs,
+    ) -> Result<PushIngestResult, PushIngestError> {
+        args.listener
+            .on_stage_progress(PushIngestStage::CheckSource, 0, TotalSteps::Exact(1));
+        args.listener
+            .on_stage_progress(PushIngestStage::Fetch, 0, TotalSteps::Exact(1));
+        args.listener
+            .on_stage_progress(PushIngestStage::Read, 0, TotalSteps::Exact(1));
+
+        let df = if let Some(df) = self.read(&args).await? {
+            if let Some(transform) = args.polling_source.preprocess.clone() {
+                Some(ingest_common::preprocess(&args.ctx, transform, df).await?)
+            } else {
+                Some(df)
+            }
+        } else {
+            tracing::info!("Read produced an empty data frame");
+            None
+        };
+
+        let out_dir = args.operation_dir.join("out");
+        let data_staging_path = out_dir.join("data");
+        std::fs::create_dir(&out_dir).int_err()?;
+
+        let stage_result = args
+            .data_writer
+            .stage(
+                df,
+                WriteDataOpts {
+                    system_time: args.system_time.clone(),
+                    source_event_time: args.system_time.clone(),
+                    source_state: None,
+                    data_staging_path,
+                },
+            )
+            .await;
+
+        match stage_result {
+            Ok(staged) => {
+                args.listener
+                    .on_stage_progress(PushIngestStage::Commit, 0, TotalSteps::Exact(1));
+
+                let res = args.data_writer.commit(staged).await?;
+
+                Ok(PushIngestResult::Updated {
+                    old_head: res.old_head,
+                    new_head: res.new_head,
+                    num_blocks: 1,
+                })
+            }
+            Err(StageDataError::BadInputSchema(e)) => Err(e.into()),
+            Err(StageDataError::MergeError(e)) => Err(e.into()),
+            Err(StageDataError::EmptyCommit(_)) => Ok(PushIngestResult::UpToDate),
+            Err(StageDataError::Internal(e)) => Err(e.into()),
+        }
+    }
+
+    #[tracing::instrument(level = "info", skip_all)]
+    async fn read(&self, args: &PushIngestArgs) -> Result<Option<DataFrame>, PushIngestError> {
+        // TODO: Support S3
+        let input_data_path: PathBuf = match args.url.scheme() {
+            "file" => {
+                let p = args
+                    .url
+                    .to_file_path()
+                    .map_err(|_| format!("Invalid file URL {}", args.url).int_err())?;
+
+                // TODO: In case of STDIN (/dev/fd/0) or other pipes and special device files
+                // we have to copy data into a temporary file, as DataFusion cannot read from
+                // them directly.
+                cfg_if::cfg_if! {
+                    if #[cfg(unix)] {
```

Comment:
What happens on Windows?

Reply:
Windows is an unsupported target - we just compile on it not to ruin our long-term chances for compatibility. If you run `kamu` on Windows today the non-working piping will probably be the smallest of issues there.